### PR TITLE
[monaco-textmate] Added new Syntax Coloring

### PR DESCRIPTION
**Pull request by marechal-p**
This PR adds better syntaxic coloring using Textmate grammars.

This relies on a library written in web assembly, so that the tokenization can run in the browser and still be as precise as on native IDEs, and very close in term of performances.

### CODE REVIEW

**Comment by svenefftinge** [comment]
> Converted the weird singleton ThemeService to an actual singleton instance managed by inversify.
That is on purpose, because we need the theme ASAP on loading.

File: dev-packages/application-manager/src/generator/webpack-generator.ts [hunk 1]
```diff
@@ -38,7 +39,8 @@ const { mode }  = yargs.option('mode', {
 }).argv;
 const development = mode === 'development';${this.ifMonaco(() => `
 
-const monacoEditorPath = development ? '${this.resolve('monaco-editor-core', 'dev/vs')}' : '${this.resolve('monaco-editor-core', 'min/vs')}';
+const monacoEditorPath = development ? '${this.resolve('monaco-editor', 'dev/vs')}' : '${this.resolve('monaco-editor', 'esm/vs')}';
```
Comment (svenefftinge) [inline dev-packages/application-manager/src/generator/webpack-generator.ts]: Why did you add monaco-editor again? Is it needed?

File: examples/browser/package.json [hunk 2]
```diff
@@ -21,6 +21,7 @@
     "@theia/metrics": "^0.3.11",
     "@theia/mini-browser": "^0.3.11",
     "@theia/monaco": "^0.3.11",
+    "@theia/monaco-textmate": "^0.3.11",
```
Comment (svenefftinge) [inline examples/browser/package.json]: I think the textmate stuff could go into the monaco extension.

File: packages/core/src/browser/browser.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@
 
 const userAgent = typeof navigator !== 'undefined' ? navigator.userAgent : "";
 
-export const isIE = (userAgent.indexOf('Trident') >= 0);
+export const isIE: boolean = (userAgent.indexOf('Trident') >= 0);
```
Comment (svenefftinge) [inline packages/core/src/browser/browser.ts]: Why adding the type here explicitly?

File: packages/core/src/browser/frontend-application-module.ts [hunk 4]
```diff
@@ -149,7 +149,13 @@ export const frontendApplicationModule = new ContainerModule((bind, unbind, isBo
         const connection = ctx.container.get(WebSocketConnectionProvider);
         return connection.createProxy<EnvVariablesServer>(envVariablesPath);
     }).inSingletonScope();
-});
 
-const theme = ThemeService.get().getCurrentTheme().id;
-ThemeService.get().setCurrentTheme(theme);
+    bind(ThemeService).toSelf().inSingletonScope();
+    bind(DefaultThemeName).toConstantValue('dark');
```
Comment (svenefftinge) [inline packages/core/src/browser/frontend-application-module.ts]: This is too late. The Theme needs to be initialized on load.

File: packages/core/src/common/theming-protocol.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright (C) 2018 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+export interface Theme {
```
Comment (svenefftinge) [inline packages/core/src/common/theming-protocol.ts]: I don't see a reason for a protocol (frontend <-> backend) communication. Also those objects need to be JSON (i.e. no functions).

File: packages/monaco-textmate/data/grammars/javascript.tmlanguage.json [hunk 6]
```diff
@@ -0,0 +1,516 @@
+{
+  "comment": "JavaScript Syntax: version 2.0",
```
Comment (svenefftinge) [inline packages/monaco-textmate/data/grammars/javascript.tmlanguage.json]: grammars belong into the respective language extensions. I.e. JS & TS into our typescript extension.

File: packages/monaco/src/browser/monaco-theme-service.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+import { injectable, inject, named } from "inversify";
+import { ContributionProvider } from "@theia/core";
+import { ThemeProvider } from "@theia/core/lib/common/theming-protocol";
+import { MonacoThemeProvider, MonacoTheme } from "../common/monaco-theme-protocol";
+import { Deferred } from "@theia/core/lib/common/promise-util";
+
+/**
+ * This class hooks onto the Theia Timport { error } from 'shelljs';
+heme service as a Theme provider, while it doesn't provide any theme.
```
Comment (svenefftinge) [inline packages/monaco/src/browser/monaco-theme-service.ts]: looks like something got deleted by accident

File: dev-packages/application-manager/src/generator/webpack-generator.ts (hunk 1 above)
Comment (marechal-p) [reply dev-packages/application-manager/src/generator/webpack-generator.ts]: it is a left over from the various tests before, nice catch !

File: packages/core/src/browser/browser.ts (hunk 3 above)
Comment (marechal-p) [reply packages/core/src/browser/browser.ts]: Left over when I was editing files with Theia + Textmate to see type annotations, I will remove it.

File: packages/monaco-textmate/data/grammars/javascript.tmlanguage.json (hunk 6 above)
Comment (marechal-p) [reply packages/monaco-textmate/data/grammars/javascript.tmlanguage.json]: It is pretty specific to textmate though, what about languages without a proper Theia extension ?

File: examples/browser/package.json (hunk 2 above)
Comment (marechal-p) [reply examples/browser/package.json]: It could, but I felt like this textmate coloring was a feature on its own, that people could work without (and maybe just use Monarch Themes instead with the correct extension). What would be the main reason to merge it in monaco ?

File: packages/core/src/common/theming-protocol.ts (hunk 5 above)
Comment (marechal-p) [reply packages/core/src/common/theming-protocol.ts]: I was expecting themes to be loadable from disk later, like some kind of `~/.theia/themes` folder where the user could drop whatever he wanted.

Now indeed this `Theme` interface is in the wrong file.

**Comment by marechal-p** [comment]
Ok, so if we want to keep the `ThemeService` as a global variable, we lose the ability to depend on inversify's bindings. I could try to inject an inversify-bound object inside it later using `Deferred` objects maybe. Or just assume that any Theme will be registered in the global variable for every extension. I'll implement the second approach first.

File: packages/monaco/src/browser/monaco-theme-service.ts (hunk 7 above)
Comment (marechal-p) [reply packages/monaco/src/browser/monaco-theme-service.ts]: Miss click with the middle mouse button under Ubuntu, it's driving me crazy.

File: packages/monaco-textmate/data/grammars/javascript.tmlanguage.json (hunk 6 above)
Comment (svenefftinge) [reply packages/monaco-textmate/data/grammars/javascript.tmlanguage.json]: Textmate would become part of the language extension API. 
But for the many textmate grammars for which we don't have extensions, we could have a bulk extension including and registering them.

**Comment by svenefftinge** [comment]
I tried it. Looks beautiful :)

Things I noticed (I know it's WIP so you are probably aware):
1) There seems to be multiple passes going on when I first open a file, as it is flickering.
2) The multi-line comments are not properly highlighted
3) Themes are not Textmate specific. Instead we should build proper themes per Textmate theme.
    For now we should use the VSCode's Dark+ and Light+ for our dark and light theme. Other themes can be created later.

The changes you made to the ThemeService seem to work nicely.

File: packages/monaco-textmate/data/grammars/javascript.tmlanguage.json (hunk 6 above)
Comment (marechal-p) [reply packages/monaco-textmate/data/grammars/javascript.tmlanguage.json]: There is a [package](https://github.com/NeekSandhu/rink-languages) that provides a lot of grammars already. Maybe this would do the trick ?

**Comment by svenefftinge** [comment]
@marechal-p let me know when I should review

**Comment by marechal-p** [comment]
@svenefftinge I fixed most of the issue mentionned, apart from the contribution point in the different language extensions. Do you see anything else ?

**Comment by svenefftinge** [comment]
Sorry, if it wasn't clear, but regarding the theme's I meant that we should just use the two textmate themes (Dark+, Light+) in the existing themes. I.e. they should be displayed as now:

screen shot 2018-06-19 at 18 27 16

The 'textmate' part is only about the syntax coloring and is an implementation detail. Not important to expose. The `+` part is a legacy suffix from VS Code. No reason for us to have it.

**Comment by svenefftinge** [comment]
The opening and closing part of a multiline comment in TS is colored bogusly (white):

screen shot 2018-06-19 at 18 30 34

**Comment by svenefftinge** [comment]
Can we get rid of `monaco-textmate` and move the code to the monaco extension?

**Comment by NeekSandhu** [comment]
@svenefftinge https://github.com/Microsoft/monaco-editor/issues/929

That's the only *clean* way out

File: packages/monaco-textmate/package.json [hunk 8]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "@theia/monaco-textmate",
+  "version": "0.3.11",
+  "description": "Theia - Monaco Extension",
+  "dependencies": {
+    "@theia/core": "^0.3.11",
+    "@theia/monaco": "^0.3.11",
+    "monaco-editor-textmate": "^2.0.0",
+    "monaco-textmate": "^3.0.0",
+    "onigasm": "^2.0.0"
```
Comment (NeekSandhu) [inline packages/monaco-textmate/package.json]: `onigasm@2.1` has huge perf improvements

Comment (NeekSandhu) [reply packages/monaco-textmate/package.json]: PS: Since this package is in evolution stage I'd recommend to keep it updated

**Comment by marechal-p** [comment]
@svenefftinge this Textmate support also entirely relies on WebASM. Is this an issue when it will affect compatibility with a good lot of browsers ? (my ESR Firefox for instance doesn't support wasm).

**Comment by marcdumais-work** [comment]
> my ESR Firefox for instance doesn't support wasm

Does it work with recent/latest, non-ESR Firefox?

**Comment by marechal-p** [comment]
> @svenefftinge Microsoft/monaco-editor#929
>
> That's the only clean way out

@NeekSandhu I ended skipping the last scope if it was starting with `punctuation.`, now the theme correctly applies on comments, strings and regular expressions !

**Comment by marechal-p** [comment]
@svenefftinge I finally exploded the `monaco-textmate` extension as you were suggesting. Didn't turn out to be as straightforward as I expected.

Can you tell me if you see anything now ?

File: packages/monaco/src/browser/textmate/monaco-builtin-theme-provider.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                    colors: {},
+                }
+            );
+
+            monaco.editor.defineTheme(theme.name, theme);
+        });
+    }
+
+    // tslint:disable-next-line:no-any
+    static convertVscodeToMonaco(vscodeTheme: any, monacoTheme: MonacoTheme): MonacoTheme {
```
Comment (svenefftinge) [inline packages/monaco/src/browser/textmate/monaco-builtin-theme-provider.ts]: doesn't need to be public

File: packages/monaco/src/browser/textmate/monaco-textmate-service.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    protected readonly toDispose = new DisposableCollection();
+
+    protected readonly activatedLanguages = new Set<string>();
+    protected textmateRegistry: Registry;
+
+    @inject(ContributionProvider) @named(LanguageGrammarDefinitionContribution)
+    protected readonly grammarProviders: ContributionProvider<LanguageGrammarDefinitionContribution>;
+
+    @inject(TextmateRegistry)
+    protected readonly theiaRegistry: TextmateRegistry;
```
Comment (svenefftinge) [inline packages/monaco/src/browser/textmate/monaco-textmate-service.ts]: naming the property `textmateRegistry`would make more sense to me
